mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py

```diff
@@ -55,24 +42,92 @@ def transfer(current_results):
             )
             continue
+        else:
+            transfer_results_for_project(project_id)
+
+    return project_id_list
+
+
+def transfer_results_for_project(project_id):
+    """Transfer the results for a specific project.
+
+    Download all results for a project from firebase.
+    Save results into an in-memory file.
+    Copy the results to postgres.
+    Delete results in firebase.
+
+    We are NOT using a Firebase transaction functions here anymore.
+    This has caused problems, in situations where a lot of mappers are
+    uploading results to Firebase at the same time. Basically, this is
+    due to the behaviour of Firebase Transaction function:
+
+        "If another client writes to this location
+        before the new value is successfully saved,
+        the update function is called again with the new current value,
+        and the write will be retried."
 
-        logger.info(f"{project_id}: Start transfering results")
+    Using Firebase transaction on the group level
+    has turned out to be too slow when using "normal" queries,
+    e.g. without using threading. Threading should be avoided here
+    as well to not run into unforeseen errors.
 
-        results_ref = fb_db.reference(f"v2/results/{project_id}")
+    For more details see issue #478.
+    """
+    # TODO: which exceptions can happen here?
+
+    logger.info(f"{project_id}: Start transfer results")
+    fb_db = auth.firebaseDB()
+    results_ref = fb_db.reference(f"v2/results/{project_id}")
+    results = results_ref.get()
+
+    if results is None:
+        logger.info(f"{project_id}: No results in Firebase")
+    else:
+        # First we check for new users in Firebase.
+        # The user_id is used as a key in the postgres database for the results
+        # and thus users need to be inserted before results get inserted.
+        results_user_id_list = get_user_ids_from_results(results)
+        update_data.update_user_data(results_user_id_list)
+
+        # Results are dumped into an in-memory file.
+        # This allows us to use the COPY statement to insert many
+        # results at relatively high speed.
+        results_file = results_to_file(results, project_id)
         truncate_temp_results()
+        save_results_to_postgres(results_file)
 
-        try:
-            results_ref.transaction(transfer)
-            logger.info(f"{project_id}: Transfered results to postgres")
-        except fb_db.TransactionAbortedError:
-            logger.exception(
-                f"{project_id}: Firebase transaction for "
-                f"transfering results failed to commit"
-            )
-            sentry.capture_exception()
+        # It is important here that we first insert results into postgres
+        # and then delete these results from Firebase.
+        # In case something goes wrong during the insert, results in Firebase
+        # will not get deleted.
+        delete_results_from_firebase(project_id, results)
```

Comment:
Here we have to be careful. This should only be executed if `save_results_to_postgres` has been successful. Currently we hope that this function (and the psycopg library) will raise an exception when inserting results to the db has not been successful. I do not know if there is need for other checks and how they would look like.

Reply:
To second what Matthias said, you want to make sure that you delete results only if there is no error when saving to postgres. You could use the python exception construct:
```python
try:
    save_results_to_postgres()
except SomePGError:
    log("failed")
else:
    delete_results_from_firebase()  # this is only run if there is no exception in the try block
```
See https://docs.python.org/3/tutorial/errors.html for more details. This only works if any problem in the postgres execution raises an exception, but I think this is a fairly safe assumption (reading the docs never hurts ;)

Reply:
Hey, 
yep this is a very good comment. I changed it based on your comments.

Reply:
Hmm, just to be on the safe side, would it make sense to add `except Exception:` before the `else:` block here? The way you do it now, if any exception that is not a `ForeignKeyViolation` is raised in the `try` block, it won't get handled, and the process will crash. If you add another handler, you can get some cleaner logging, and keep going. I'm not sure what makes more sense.

Reply:
yes, you are right. I added some new lines.